runtime/src/iree/builtins/ukernel/arch/x86_64/mmt4d_x86_64_avx2_fma.c

```diff
@@ -200,6 +200,81 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0(
     iree_uk_mmt4d_tile_s8s8s32_1x8x2_to_4x8x2_x86_64_avx2_fma,
     iree_uk_mmt4d_tile_s8s8s32_4x8x2_x86_64_avx2_fma, 4)
 
+IREE_UK_ATTRIBUTE_ALWAYS_INLINE static inline void
+iree_uk_mmt4d_tile_s8s4s32_1x8x4_to_8x8x4_x86_64_avx2_fma(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 8 && iree_uk_is_po2_u32(M0));
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  __m256i acc[8];
+  if (params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE) {
+    IREE_UK_UNROLL for (int i = 0; i < M0; ++i) {
+      acc[i] = _mm256_loadu_si256((__m256i*)(out_ptr + i * 8));
+    }
+  } else {
+    IREE_UK_UNROLL for (int i = 0; i < M0; ++i) {
+      acc[i] = _mm256_setzero_si256();
+    }
+  }
+
+  // Mask for extracting high nibbles.
+  __m128i low_nibble_mask = _mm_set1_epi16(0xf0f0);
+
+  for (int k = 0; k < params->K; ++k) {
+    // rhs_i4 is the rhs_tile 8x4xi4.
+    __m128i rhs_i4 = _mm_loadu_si128((const __m128i*)rhs_ptr);
+
+    // Mask low nibbles. The high values can now be represented in i8 but due to
+    // the mask the value is multipled by 2^4.
+    __m128i rhs_high_i8 = _mm_and_si128(rhs_i4, low_nibble_mask);
+    __m256i rhs_high_i16 = _mm256_cvtepi8_epi16(rhs_high_i8);
+
+    // To extract the low nibbles; we shift left by 4 bits and apply the
+    // same mask as above.
+    __m128i rhs_low_i4 = _mm_slli_epi16(rhs_i4, 4);
+    __m128i rhs_low_i8 = _mm_and_si128(rhs_low_i4, low_nibble_mask);
```

Comment:
I see. You're left-shifting also the low bits so they, too, end up occupying the high bits, so that the 16x scale factor applies to all the products and you don't need separate accumulator. Worth a comment? Because if it weren't for that idea, I would have suggested just directly getting the low bits, without a shift and using the same mask constant, using a _andnot_ intrinsic to negate the masking semantics. For instance, `_mm_andnot_si128`: https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=andnot_si128&ig_expand=343